lib/rubygems/gem_runner.rb

```diff
@@ -74,6 +82,12 @@ def do_configuration(args)
     Gem.use_paths Gem.configuration[:gemhome], Gem.configuration[:gempath]
     Gem::Command.extra_args = Gem.configuration[:gem]
   end
+
+  def warn_on_outdated
+    if defined?(OpenSSL::SSL) && (Gem.latest_rubygems_version > Gem.rubygems_version)
```

Comment:
`Gem.latest_rubygems_version` does remote request, which can fail and raise.

```ruby
irb(main):001:0> Gem.latest_rubygems_version                                                                                                                                                  
Traceback (most recent call last):
        5: from /home/retro/.rubies/ruby-2.7.2/bin/irb:23:in `<main>'
        4: from /home/retro/.rubies/ruby-2.7.2/bin/irb:23:in `load'                            
        3: from /home/retro/.rubies/ruby-2.7.2/lib/ruby/gems/2.7.0/gems/irb-1.2.6/exe/irb:11:in `<top (required)>'
        2: from (irb):1                                                                                                                                                                       
        1: from /home/retro/.rubies/ruby-2.7.2/lib/ruby/2.7.0/rubygems.rb:952:in `latest_rubygems_version'
RuntimeError (Can't find 'rubygems-update' in any repo. Check `gem source list`.)
```

With current implementation, any gem command (even `gem help`) would fail for example when offline. I think we should do this check only for various commands (like gem install, update, ...) where it is common to hit the remote. Also we should handle gracefully when `Gem.latest_rubygems_version ` fails or it doesn't respond in reasonable time.

Reply:
I really like `yarn` warning about the update. It is not annoying (I had no idea until I looked into code).

https://github.com/yarnpkg/yarn/blob/a4708b29ac74df97bac45365cba4f1d62537ceb7/src/cli/commands/install.js#L1067-L1126

1. don't show upgrade dialog on CI or non-TTY terminals
2. don't check if disabled in user config
3. only check for updates once a day
4. don't bug for updates on tagged releases (don't warn if version includes `-`, I think that makes it friendly for rubygems devs)
5. swallow errors (if anything goes wrong, just ignore and swallow all exceptions)